### ISSUE TITLE
Restore python dependency validation

### ISSUE THREAD

**Issue by tjni** [NixOS/nixpkgs#250865]
### Describe the bug

I did not realize this when I was implementing installing using [installer](https://installer.pypa.io/en/stable/) instead of pip, but installer doesn't perform any dependency validation. I think we should restore this functionality. Should we use `pip install --dry-run` to do this as a pre-install step (does that even work)? Or maybe we can parse the metadata ourselves.

### Notify maintainers

@FRidh @mweinelt @jonringer

**Comment by jonringer** [comment]
I would much rather have something in `passthru.tests` which shows that a trivial use cases succeeds, like a more thorough version of `pythonImportsCheck`. This should catch a much broader issues besides just build issues. I think issues with python's preferences toward relative directories should be exceedingly rare, and more of an exception than the rule.

However, if `pip install --dry-run` is relatively cheap (less than a second) to do, I would be for it. However, even if it's ~10 seconds, then multiply that with 5000+ packages, and the additional build times could be significant. (90+ additional minutes for every python-updates or staging run).

> Or maybe we can parse the metadata ourselves.

 I've had issues in the past where a dependency isn't included in setup.py (or other metadata file); and the package is actually broken when you package it strictly with what was declared. The packages worked locally because they usually ran multiple commands which eventually got them into a coherent state.

**Comment by FRidh** [comment]
That could be a more robust solution, however, as it is now, Hydra does not evaluate the tests attribute. We'd have to collect those attribute sets and merge them into some other attribute set Hydra would become aware of.

**Comment by FRidh** [comment]
I asked for an alternative to pip check https://discuss.python.org/t/lightweight-alternative-to-pip-check/32204

**Comment by FRidh** [comment]
I suggest we use https://github.com/pypa/build/blob/a167ba4c03bcf1b74c04d291d95fdfc48942bbd5/src/build/_util.py#L18, either by reusing the build library, or just taking that piece of code and put it in a script.

**Comment by mweinelt** [comment]
I came up with the following, and we could run it as a hook before or early in `pypaInstallPhase`. It does however require `packaging>=23.2` to the heavy-lifting.

```python
#!/usr/bin/env python3
"""
The runtimeDependenciesHook validates, that all dependencies specified
in wheel metadata are available in the local environment.

In case that does not hold, it will print missing dependencies and
violated version constraints. 
"""


import importlib.metadata
import sys
import tempfile
from argparse import ArgumentParser
from zipfile import ZipFile

from packaging.metadata import Metadata, parse_email
from packaging.requirements import Requirement

argparser = ArgumentParser()
argparser.add_argument("wheel", help="Path to the .whl file to test")


def get_manifest_text_from_wheel(wheel: str) -> str:
    """
    Given a path to a wheel, this function will try to extract the
    METADATA file in the wheels .dist-info directory.
    """
    with ZipFile(wheel) as zipfile:
        for zipinfo in zipfile.infolist():
            if zipinfo.filename.endswith(".dist-info/METADATA"):
                with tempfile.TemporaryDirectory() as tmp:
                    path = zipfile.extract(zipinfo, path=tmp)
                    with open(path, encoding="utf-8") as fd:
                        return fd.read()

    raise RuntimeError("No METADATA file found in wheel")


def get_manifest(wheel: str) -> Metadata:
    """
    Given a path to a wheel, returns a parsed Metadata object.
    """
    text = get_manifest_text_from_wheel(wheel)
    raw, _ = parse_email(text)
    metadata = Metadata.from_raw(raw)

    return metadata


def test_requirement(requirement: Requirement) -> bool:
    """
    Given a requirement specification, tests whether the dependency can
    be resolved in the local environment, and whether it satisfies the
    specified version constraints.
    """
    if requirement.marker and not requirement.marker.evaluate():
        return True

    try:
        package = importlib.metadata.distribution(requirement.name)
    except importlib.metadata.PackageNotFoundError:
        print(f"- {requirement} missing")
        return False

    if package.version not in requirement.specifier:
        print(
            f"- {package.name}=={package.version} does not satisfy {requirement.specifier}",
            file=sys.stderr,
        )
        return False

    return True


if __name__ == "__main__":
    args = argparser.parse_args()

    manifest = get_manifest(args.wheel)
    satisfied = all(
        test_requirement(requirement) for requirement in manifest.requires_dist
    )

    if not satisfied:
        sys.exit(1)
```

**Comment by mweinelt** [comment]
Fixed in #270457, which went into #271586 and #275620.
